OAK-9565: NullPointerException on Session.logout()

mreutegg opened the pull request:
Test case with proposed fix.

mreutegg commented:
`SessionStatsTest.testInitStackTraceEnabledAfterOpeningManySessions` fails because of a side effect of the new `SessionLogoutTest`. The new test provokes the issue described in OAK-9262. The initStackTrace feature and the test depends on a correct session counter, which is violated when `SessionLogoutTest` runs.

I'm going to change the fix to also include OAK-9262.

stefan-egli commented on oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/session/SessionLogoutTest.java:
```diff
@@ -0,0 +1,56 @@
+
+    public void testConcurrentSessionLogout() throws Exception {
+        Session s = getHelper().getReadWriteSession();
+        List<Exception> exceptions = new ArrayList<>();
```
(Just a note to self: I was first wondering whether you'd need a concurrent version of List (eg ConcurrentLinkedQueue) - but Thread.join indeed, and somewhat obviously, honours the ['happens-before order' ](https://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.4.5), so this works perfectly fine.)

mreutegg replied:
Hmm, right. But the List must still be thread-safe, which is currently not the case. Multiple threads may add exceptions to the list concurrently in the failure case. I'm going to change that.

mreutegg committed:
OAK-9565: NullPointerException on Session.logout()

SessionLogoutTest to reproduce the issue
Also introduce test from OAK-9262
Synchronize Session.logout() on a newly introduced monitor object
